src/test/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepDynamicContextRJRTest.java

```diff
@@ -0,0 +1,127 @@
+import org.jvnet.hudson.test.RealJenkinsRule;
+import org.jvnet.hudson.test.TailLog;
+
+public class ExecutorStepDynamicContextRJRTest {
+    private static final Logger LOGGER = Logger.getLogger(ExecutorStepDynamicContextRJRTest.class.getName());
+
+    @Rule public RealJenkinsRule rjr = new RealJenkinsRule().withColor(PrefixedOutputStream.Color.GREEN).withPackageLogger(ExecutorStepExecution.class, Level.FINE).withPackageLogger(FlowExecutionList.class, Level.FINE);
+
+    @Rule
+    public InboundAgentRule iar = new InboundAgentRule();
+
+    @Test public void restartWhileWaitingForANode() throws Throwable {
+        rjr.startJenkins();
+        try (var tailLog = new TailLog(rjr, "p", 1)) {
+            iar.createAgent(rjr, InboundAgentRule.Options.newBuilder().name("J").label("mib").color(PrefixedOutputStream.Color.YELLOW).webSocket().build());
+            rjr.runRemotely(ExecutorStepDynamicContextRJRTest::setupJobAndStart);
+            rjr.stopJenkinsForcibly();
+            rjr.startJenkins();
+            rjr.runRemotely(ExecutorStepDynamicContextRJRTest::assertQueueItems);
+        }
+    }
+
+    private static void assertQueueItems(JenkinsRule r) throws Throwable {
+        var p = r.jenkins.getItemByFullName("p", WorkflowJob.class);
+        var b = p.getBuildByNumber(1);
+        var items = await("Waiting for agent J to reconnect").atMost(Duration.ofSeconds(30)).until(() -> r.jenkins.getQueue().getItems(), arrayWithSize(1));
+        LOGGER.info("Items in queue: " + Arrays.asList(items));
+        var actions = await().until(() -> b.getActions(InputAction.class), hasSize(1));
+        proceed(actions, "Complete branch 1 ?", p.getName() + "#" + b.number);
+        actions = await().until(() -> b.getActions(InputAction.class), allOf(iterableWithSize(1), hasItem(new InputActionIsWaitingForInput())));
+        r.waitForMessage("Complete branch 2 ?", b);
+        proceed(actions, "Complete branch 2 ?", p.getName() + "#" + b.number);
+        r.waitForCompletion(b);
+    }
+
+    private static class InputActionIsWaitingForInput extends TypeSafeMatcher<InputAction> {
+
+        @Override
+        protected boolean matchesSafely(InputAction inputAction) {
+            try {
+                return inputAction.isWaitingForInput();
+            } catch (Exception e) {
+                return false;
+            }
+        }
+
+        @Override
+        public void describeTo(Description description) {
+            description.appendText("is waiting for input");
+        }
+    }
+
+    private static void proceed(List<InputAction> actions, String inputMessage, String name) throws InterruptedException, TimeoutException, IOException {
+        for (var action : actions) {
+            if (action.getExecutions().isEmpty()) {
+                continue;
+            }
+            var inputStepExecution = action.getExecutions().get(0);
+            if (inputMessage.equals(inputStepExecution.getDisplayName())) {
```

Comment:
Clearer to use an `id`. (Just beware that they are forcibly capitalized, so use a capitalized id to start with.)